# Preprocessing the model: Applying Feature Engineering to the Dataset

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

## Loading the data

In [2]:
root = '../data/cleaned_cars.csv'

df = pd.read_csv(root)
df.head()

title       brand     model    type  year  \
0  Smart Fortwo Cabrio 52 Mhd Pure Aut.       SMART    FORTWO  sports  2014   
1           Volkswagen Scirocco 2.0 Tsi  VOLKSWAGEN  SCIROCCO  sports  2009   
2          Bmw Serie 2 218da Gran Coupé         BMW   SERIE 2  sports  2021   
3                     Bmw X4 Xdrive 20d         BMW        X4  sports  2017   
4                    Dodge Viper Srt-10       DODGE     VIPER  sports  2005   

     kms       city    gearbox  doors  seats  ...  width trunk_vol  max_speed  \
0  37125   Alicante  Automatic      2      2  ...    156       203        145   
1      0  Barcelona     Manual      3      4  ...    181       292        235   
2      0  Cantabria  Automatic      4      5  ...    180       430        222   
3  85000     Girona  Automatic      5      5  ...    188       500        212   
4  95017    Unknown  Automatic      2      2  ...    191       344        189   

  urban_cons xtrurban_cons mixed_cons weight  tank_vol   acc  price  
0        4.6           4.0        4.3    780        33  13.7   5500  
1        9.9           6.1        7.6   1373        55   7.2  10900  
2        5.4           2.9        4.2   1545        42   8.5  36100  
3        6.1           5.0        5.4   1740        67   8.0  28000  
4       27.3          14.7       21.0   1546        70   3.9  27990  

[5 rows x 29 columns]

## Identifying categorical features

In [3]:
df.select_dtypes(exclude=["number"]).columns.value_counts().reset_index().drop(0, axis=1)

index
0     gearbox
1       brand
2      dealer
3        city
4        type
5       title
6   fuel_type
7       color
8       model
9    warranty
10    chassis

In [4]:
df.select_dtypes(exclude=["number"])

title       brand  \
0                   Smart Fortwo Cabrio 52 Mhd Pure Aut.       SMART   
1                            Volkswagen Scirocco 2.0 Tsi  VOLKSWAGEN   
2                           Bmw Serie 2 218da Gran Coupé         BMW   
3                                      Bmw X4 Xdrive 20d         BMW   
4                                     Dodge Viper Srt-10       DODGE   
...                                                  ...         ...   
55717                    Opel Corsa 1.3cdti Selective 75        OPEL   
55718   Mercedes Clase V 250d Largo Avantgarde 7g Tronic    MERCEDES   
55719             Jeep Cherokee 2.0d Longitude 4x2 103kw        JEEP   
55720                                 Bmw X3 Xdrive 30da         BMW   
55721  Citroen C4 Aircross 1.6hdi S&s Exclusive Plus ...     CITROEN   

          model    type         city    gearbox  color fuel_type warranty  \
0        FORTWO  sports     Alicante  Automatic  WHITE  Gasoline      YES   
1      SCIROCCO  sports    Barcelona     Manual  WHITE  Gasoline      YES   
2       SERIE 2  sports    Cantabria  Automatic  OTHER    Diesel      YES   
3            X4  sports       Girona  Automatic  WHITE    Diesel      YES   
4         VIPER  sports      Unknown  Automatic    RED  Gasoline      YES   
...         ...     ...          ...        ...    ...       ...      ...   
55717     CORSA   other       Madrid     Manual  WHITE    Diesel      YES   
55718   CLASE V   other      Unknown  Automatic  BLACK    Diesel      YES   
55719  CHEROKEE   other  Ciudad Real     Manual   GREY    Diesel      YES   
55720        X3   other    Barcelona  Automatic  WHITE    Diesel      YES   
55721        C4   other      Sevilla     Manual   GREY    Diesel      YES   

             dealer      chassis  
0      Professional  Convertible  
1      Professional        Coupe  
2      Professional        Coupe  
3      Professional        Coupe  
4      Professional        Coupe  
...             ...          ...  
55717  Professional        Sedan  
55718  Professional      Minivan  
55719  Professional      Offroad  
55720  Professional      Offroad  
55721  Professional      Offroad  

[55722 rows x 11 columns]

It is needed to convert these categorical variables to numbers using a OneHotEncoder. However, title will be the exception since it is not valuable to include long text in a regression model.

In [5]:
cat_cols = df.select_dtypes(exclude=["number"]).columns.drop('title')

### Transforming Categorical into Numerical: One Hot Encoding

Using a dummifyer (`pd.get_dummies`) to transform categorical values into numerical is not the right approach for a machine learning tool. `LabelEncoder` will interpret a categorical feature i.e. `[a,b,b,c]` as `[0,1,1,2]`. This will establish a wrong numeric relationship among the categorical variable. Therefore, using a `OneHotEncoder` is the right procedure to represent a categorical variable since it will create a new category per value.  

In [6]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelBinarizer

### When should we use each preprocessing encoding functions?

- **LabelEncoder** - for labels(response variable) coding 1,2,3… [implies order]

- **OrdinalEncoder** - for features coding 1,2,3 … [implies order]

- **Label Binarizer** - for response variable, coding 0 & 1 [ creating multiple dummy columns]

- **OneHotEncoder** - for feature variables, coding 0 & 1 [ creating multiple dummy columns]

In [7]:
onehot_encoder = LabelBinarizer()

In [8]:
df.select_dtypes(exclude=["number"])

title       brand  \
0                   Smart Fortwo Cabrio 52 Mhd Pure Aut.       SMART   
1                            Volkswagen Scirocco 2.0 Tsi  VOLKSWAGEN   
2                           Bmw Serie 2 218da Gran Coupé         BMW   
3                                      Bmw X4 Xdrive 20d         BMW   
4                                     Dodge Viper Srt-10       DODGE   
...                                                  ...         ...   
55717                    Opel Corsa 1.3cdti Selective 75        OPEL   
55718   Mercedes Clase V 250d Largo Avantgarde 7g Tronic    MERCEDES   
55719             Jeep Cherokee 2.0d Longitude 4x2 103kw        JEEP   
55720                                 Bmw X3 Xdrive 30da         BMW   
55721  Citroen C4 Aircross 1.6hdi S&s Exclusive Plus ...     CITROEN   

          model    type         city    gearbox  color fuel_type warranty  \
0        FORTWO  sports     Alicante  Automatic  WHITE  Gasoline      YES   
1      SCIROCCO  sports    Barcelona     Manual  WHITE  Gasoline      YES   
2       SERIE 2  sports    Cantabria  Automatic  OTHER    Diesel      YES   
3            X4  sports       Girona  Automatic  WHITE    Diesel      YES   
4         VIPER  sports      Unknown  Automatic    RED  Gasoline      YES   
...         ...     ...          ...        ...    ...       ...      ...   
55717     CORSA   other       Madrid     Manual  WHITE    Diesel      YES   
55718   CLASE V   other      Unknown  Automatic  BLACK    Diesel      YES   
55719  CHEROKEE   other  Ciudad Real     Manual   GREY    Diesel      YES   
55720        X3   other    Barcelona  Automatic  WHITE    Diesel      YES   
55721        C4   other      Sevilla     Manual   GREY    Diesel      YES   

             dealer      chassis  
0      Professional  Convertible  
1      Professional        Coupe  
2      Professional        Coupe  
3      Professional        Coupe  
4      Professional        Coupe  
...             ...          ...  
55717  Professional        Sedan  
55718  Professional      Minivan  
55719  Professional      Offroad  
55720  Professional      Offroad  
55721  Professional      Offroad  

[55722 rows x 11 columns]

In [17]:
onehot_encoded = onehot_encoder.fit_transform(df.type)
df_ohe = pd.DataFrame(onehot_encoded, columns=np.unique(df.type.values))
df_ohe.head()

big  familiar  medium  minivan  offroad  other  small  sports  van
0    0         0       0        0        0      0      0       1    0
1    0         0       0        0        0      0      0       1    0
2    0         0       0        0        0      0      0       1    0
3    0         0       0        0        0      0      0       1    0
4    0         0       0        0        0      0      0       1    0

In [18]:
data = pd.concat([df, df_ohe], axis=1)
data.head()

title       brand     model    type  year  \
0  Smart Fortwo Cabrio 52 Mhd Pure Aut.       SMART    FORTWO  sports  2014   
1           Volkswagen Scirocco 2.0 Tsi  VOLKSWAGEN  SCIROCCO  sports  2009   
2          Bmw Serie 2 218da Gran Coupé         BMW   SERIE 2  sports  2021   
3                     Bmw X4 Xdrive 20d         BMW        X4  sports  2017   
4                    Dodge Viper Srt-10       DODGE     VIPER  sports  2005   

     kms       city    gearbox  doors  seats  ...  price big  familiar medium  \
0  37125   Alicante  Automatic      2      2  ...   5500   0         0      0   
1      0  Barcelona     Manual      3      4  ...  10900   0         0      0   
2      0  Cantabria  Automatic      4      5  ...  36100   0         0      0   
3  85000     Girona  Automatic      5      5  ...  28000   0         0      0   
4  95017    Unknown  Automatic      2      2  ...  27990   0         0      0   

  minivan offroad other  small  sports  van  
0       0       0     0      0       1    0  
1       0       0     0      0       1    0  
2       0       0     0      0       1    0  
3       0       0     0      0       1    0  
4       0       0     0      0       1    0  

[5 rows x 38 columns]

In [19]:
df.type.value_counts().reset_index().sort_values('index')

index  type
1       big  9736
6  familiar  3028
0    medium  9807
3   minivan  8228
2   offroad  9551
8     other  1659
7     small  2729
4    sports  5524
5       van  5460

Let's verify if the OneHotEncoder was correctly applied:

In [20]:
for col in np.unique(df.type):
    print('N {} = {}'.format(col, data[col].sum()))

N big = 9736
N familiar = 3028
N medium = 9807
N minivan = 8228
N offroad = 9551
N other = 1659
N small = 2729
N sports = 5524
N van = 5460


The One Hot Encoder was succesfully applied to the model:

In [21]:
for col in np.unique(df.type):
    print(data[data.type == col][['type', col]].head())

     type  big
5524  big    1
5525  big    1
5526  big    1
5527  big    1
5528  big    1
           type  familiar
51035  familiar         1
51036  familiar         1
51037  familiar         1
51038  familiar         1
51039  familiar         1
         type  medium
41228  medium       1
41229  medium       1
41230  medium       1
41231  medium       1
41232  medium       1
          type  minivan
15260  minivan        1
15261  minivan        1
15262  minivan        1
15263  minivan        1
15264  minivan        1
          type  offroad
31677  offroad        1
31678  offroad        1
31679  offroad        1
31680  offroad        1
31681  offroad        1
        type  other
54063  other      1
54064  other      1
54065  other      1
54066  other      1
54067  other      1
        type  small
28948  small      1
28949  small      1
28950  small      1
28951  small      1
28952  small      1
     type  sports
0  sports       1
1  sports       1
2  sports       1
3  sports       1
4  s

Therefore, we can drop now the `type` column

In [22]:
data = data.drop('type', axis=1)
data.head()

title       brand     model  year    kms  \
0  Smart Fortwo Cabrio 52 Mhd Pure Aut.       SMART    FORTWO  2014  37125   
1           Volkswagen Scirocco 2.0 Tsi  VOLKSWAGEN  SCIROCCO  2009      0   
2          Bmw Serie 2 218da Gran Coupé         BMW   SERIE 2  2021      0   
3                     Bmw X4 Xdrive 20d         BMW        X4  2017  85000   
4                    Dodge Viper Srt-10       DODGE     VIPER  2005  95017   

        city    gearbox  doors  seats  power  ...  price  big familiar medium  \
0   Alicante  Automatic      2      2     71  ...   5500    0        0      0   
1  Barcelona     Manual      3      4    200  ...  10900    0        0      0   
2  Cantabria  Automatic      4      5    150  ...  36100    0        0      0   
3     Girona  Automatic      5      5    190  ...  28000    0        0      0   
4    Unknown  Automatic      2      2    506  ...  27990    0        0      0   

  minivan offroad  other  small  sports  van  
0       0       0      0      0       1    0  
1       0       0      0      0       1    0  
2       0       0      0      0       1    0  
3       0       0      0      0       1    0  
4       0       0      0      0       1    0  

[5 rows x 37 columns]

It is needed to build a module to deal with all the categorical columns and transform them into numerical.

Therefore, to apply these changes to all the columns, pd.get_dummies will be used.

In [30]:
df.select_dtypes(include=["object"]).drop(['title'], axis=1).columns

Index(['brand', 'model', 'type', 'city', 'gearbox', 'color', 'fuel_type',
       'warranty', 'dealer', 'chassis'],
      dtype='object')

In [31]:
cat_cols = df.select_dtypes(include=["object"]).drop(['title'], axis=1).columns
df_dummy = pd.get_dummies(df, columns=cat_cols)
df_dummy.head()

title  year    kms  doors  seats  power  \
0  Smart Fortwo Cabrio 52 Mhd Pure Aut.  2014  37125      2      2     71   
1           Volkswagen Scirocco 2.0 Tsi  2009      0      3      4    200   
2          Bmw Serie 2 218da Gran Coupé  2021      0      4      5    150   
3                     Bmw X4 Xdrive 20d  2017  85000      5      5    190   
4                    Dodge Viper Srt-10  2005  95017      2      2    506   

   co2_emiss  height  length  width  ...  chassis_Convertible  chassis_Coupe  \
0         99     157     270    156  ...                    1              0   
1        179     140     426    181  ...                    0              1   
2        109     142     453    180  ...                    0              1   
3        142     162     467    188  ...                    0              1   
4        488     123     446    191  ...                    0              1   

   chassis_Minivan  chassis_Offroad  chassis_Pickup  chassis_Roadster  \
0                0                0               0                 0   
1                0                0               0                 0   
2                0                0               0                 0   
3                0                0               0                 0   
4                0                0               0                 0   

   chassis_Sedan  chassis_Stationwagon  chassis_Targa  chassis_Van  
0              0                     0              0            0  
1              0                     0              0            0  
2              0                     0              0            0  
3              0                     0              0            0  
4              0                     0              0            0  

[5 rows x 896 columns]

In [34]:
print("Initial number of features:", df.shape[1])
print("Final number of features:", df_dummy.shape[1])

Initial number of features: 29
Final number of features: 896


The DataFrame was transformed to get `896 features` from the initial **29 features**

---

## Exploring Numerical features

On the other hand, there are more numerical columns that will be included in the regression model

In [23]:
df.select_dtypes(exclude=["object"]).columns.value_counts().reset_index().drop(0, axis=1)

index
0       max_speed
1             acc
2        tank_vol
3           price
4   xtrurban_cons
5           doors
6            year
7           seats
8           width
9       co2_emiss
10     mixed_cons
11      trunk_vol
12         weight
13         height
14          power
15     urban_cons
16            kms
17         length

In [24]:
df.select_dtypes(exclude=["object"])

year     kms  doors  seats  power  co2_emiss  height  length  width  \
0      2014   37125      2      2     71         99     157     270    156   
1      2009       0      3      4    200        179     140     426    181   
2      2021       0      4      5    150        109     142     453    180   
3      2017   85000      5      5    190        142     162     467    188   
4      2005   95017      2      2    506        488     123     446    191   
...     ...     ...    ...    ...    ...        ...     ...     ...    ...   
55717  2016   60000      5      5     75        100     148     402    175   
55718  2016   70261      5      6    190        158     188     514    193   
55719  2015  182069      5      5    140        139     163     462    186   
55720  2011  159000      5      5    258        149     166     465    188   
55721  2016   78000      5      5    114        132     163     434    180   

       trunk_vol  max_speed  urban_cons  xtrurban_cons  mixed_cons  weight  \
0            203        145         4.6            4.0         4.3     780   
1            292        235         9.9            6.1         7.6    1373   
2            430        222         5.4            2.9         4.2    1545   
3            500        212         6.1            5.0         5.4    1740   
4            344        189        27.3           14.7        21.0    1546   
...          ...        ...         ...            ...         ...     ...   
55717        285        164         4.6            3.3         3.8    1237   
55718       1030        206         6.9            5.5         6.0    2145   
55719        412        187         6.4            4.6         5.3    1828   
55720        550        230         7.8            4.8         6.0    1800   
55721        419        180         5.6            4.7         5.0    1500   

       tank_vol   acc  price  
0            33  13.7   5500  
1            55   7.2  10900  
2            42   8.5  36100  
3            67   8.0  28000  
4            70   3.9  27990  
...         ...   ...    ...  
55717        45  14.8   7500  
55718        67   9.1  48500  
55719        60  10.9  15999  
55720        67   6.2  18300  
55721        60  11.6  15500  

[55722 rows x 18 columns]